litestar/dto/factory/_backends/utils.py

```diff
@@ -129,20 +112,150 @@ def _camelize(string: str, capitalize_first_letter: bool = False) -> str:
+        transfer_type = field_definition.transfer_type
+        source_name = (
+            field_definition.serialization_name or field_definition.name if dto_for == "data" else field_definition.name
+        )
+        destination_name = (
+            field_definition.name if dto_for == "data" else field_definition.serialization_name or field_definition.name
+        )
+        source_value = getattr(source_instance, source_name)
+        if isinstance(transfer_type, SimpleType) and transfer_type.transfer_model:
+            unstructured_data[destination_name] = _transfer_instance_data(
+                transfer_type.parsed_type.annotation if dto_for == "data" else transfer_type.transfer_model.model,
+                source_value,
+                transfer_type.transfer_model.field_definitions,
+                dto_for,
+            )
+        elif isinstance(transfer_type, UnionType) and transfer_type.has_nested:
+            for inner_type in transfer_type.inner_types:
+                if (
+                    isinstance(inner_type, SimpleType)
+                    and inner_type.transfer_model
+                    and isinstance(
+                        source_value,
+                        inner_type.transfer_model.model if dto_for == "data" else inner_type.parsed_type.annotation,
+                    )
+                ):
+                    unstructured_data[destination_name] = _transfer_instance_data(
+                        inner_type.parsed_type.annotation if dto_for == "data" else inner_type.transfer_model.model,
+                        source_value,
+                        inner_type.transfer_model.field_definitions,
+                        dto_for,
+                    )
+        elif isinstance(transfer_type, CollectionType) and transfer_type.has_nested:
+            if field_definition.parsed_type.origin is None:  # pragma: no cover
+                raise RuntimeError("Unexpected origin value for collection type.")
+
+            if not isinstance(transfer_type.inner_type, SimpleType):
+                raise RuntimeError("Compound inner types not yet supported")
+
+            if transfer_type.inner_type.transfer_model is None:
+                raise RuntimeError("Inner type expected to have transfer model")
+
+            dest_type = (
+                transfer_type.inner_type.parsed_type.annotation
+                if dto_for == "data"
+                else transfer_type.inner_type.transfer_model.model
+            )
+            unstructured_data[destination_name] = field_definition.parsed_type.origin(
+                _transfer_instance_data(
+                    dest_type, item, transfer_type.inner_type.transfer_model.field_definitions, dto_for
+                )
+                for item in source_value
+            )
+        else:
+            unstructured_data[destination_name] = source_value
+
+    return destination_type(**unstructured_data)
+
+
+def _transfer_data(
```

Comment:
remove leading underscore - as this whole module is within a private sub-package, all private functions in this module are unnecessary - api can be controlled with `__all__`.